core/cache/unit.go

```diff
@@ -66,9 +70,58 @@ func (u *Unit) Ports() []network.Port {
 	return u.details.Ports
 }
 
-// WatchCharmConfig returns a new watcher that will notify when the
+// Config settings returns the effective charm configuration for this unit
+// taking into account whether it is tracking a model branch.
+func (u *Unit) ConfigSettings() (charm.Settings, error) {
+	appName := u.details.Application
+	app, err := u.model.Application(appName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cfg := app.Config()
+	if cfg == nil {
+		cfg = make(map[string]interface{})
+	}
+
+	// Apply any branch-based deltas to the master settings
+	var deltas settings.ItemChanges
+	for _, b := range u.model.Branches() {
+		if units := b.AssignedUnits()[appName]; len(units) > 0 {
+			if set.NewStrings(units...).Contains(u.details.Name) {
+				deltas = b.AppConfig(appName)
+				break
+			}
+		}
+	}
+
+	for _, delta := range deltas {
+		switch {
```

Comment:
This pattern is repeated in a few places, but is set to be modularised in a forthcoming patch.